store/postgres/migrations/2019-07-20-195916_remove_subgraph/up.sql

```diff
@@ -0,0 +1,122 @@
+-- List all metadata entities for a set of subgraph deployments.
+-- The entities are returned as a table of (id, entity) pairs, in no particular
+-- order. The list contains all the entities that are reachable through
+-- associations from a SubgraphDeployment, and include the SubgraphDeployment
+-- itself.
+create or replace function
+  list_deployment_entities(ids varchar[])
+returns table(id varchar, entity varchar) as
+$$
+begin
+  create temp table tmp_ids
+      on commit drop as
+  select * from unnest(ids) as id;
+  create index tmp_ids_id on tmp_ids(id);
+
+  -- Put the ids of all dynamic data sources for the
+  -- subgraph deployments in ids into tmp_dds
+  create temp table tmp_dds
+      on commit drop as
+  select dds.id
+    from subgraphs.entities dds
+   where dds.entity = 'DynamicEthereumContractDataSource'
+     and dds.data->'deployment'->>'data' = any(ids);
+   create index tmp_dds_id on tmp_dds(id);
+
+   -- The metadata entities for a subgraph deployment consist of two
+   -- distinct groups of entities:
+   --   * static metadata, whose id starts with the id of the
+   --     subgraph deployment. Subgraph deployment ids are 46 chars long
+   --   * metadata for dynamic data sources, whose ids start with
+   --     the id of the dynamic data source. Dynamic data source ids are
+   --     40 chars long
+   return query
+   select e.id, e.entity
+     from subgraphs.entities e, tmp_ids
+    where substring(e.id, 1, 46) = tmp_ids.id
+   union all
+   select e.id, e.entity
+     from subgraphs.entities e, tmp_dds
+    where substring(e.id, 1, 40) = tmp_dds.id;
+end;
+$$ language plpgsql;
+
+-- Remove the metadata for the deployments with the given ids. None of the
+-- deployments must currently be assigned to a node for indexing.
+-- Returns the number of entities removed.
+-- Note that this function only removes the metadata from
+-- subgraphs.entities, but not the actual data for the subgraph
+create or replace function
+  remove_deployment_metadata(ids varchar[])
+returns int as
+$$
+declare
+  rows_deleted int;
+begin
+  if exists (select 1
+               from subgraphs.entities
+              where entity = 'SubgraphDeploymentAssignment'
+                and id = any(ids)) then
+     raise 'Can not remove assigned subgraph. One of % is currently assigned',
+           ids;
+  end if;
+
+  drop table if exists tmp_deployments;
```

Comment:
Why have this drop for `tmp_deployments` but not for `tmp_ids` and `tmp_dds`?

Reply:
IIRC, I put that in when I was testing it. I am a little loathe to remove it since I'd need to retest, and it won't cause any harm there.